contracts/PreOrder.sol

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./BatchNFT.sol";
+
+contract PreOrder is BatchNFT {
+    struct Participant {
+        address sender;
+        uint256 amount;
+        uint256 time;
+    }
+
+    // the key `_nextPreOrder` is incremented with each additional pre-order
+    // make sure `_nextPreOrder` starts from 1 instead of 0
+    mapping(address => uint256) private _preOrdered;
+    mapping(uint256 => Participant) private _preOrders;
+    uint256 private _nextPreOrder;
+    uint256 private _preOrderAmountTotal;
+
+    // determine whether there is an ongoing pre-order
+    bool public inPreOrder;
+    // unit (wei)
+    uint256 public minAmount;
+    // participants allowed
+    uint256 public participantsAllowed;
+
+    // error PreOrderNotStarted();
+    // error PreOrderAlreadyStarted();
+    // error PreOrderExists(address addr);
+    // minimum contribution amount allowed shall not be <= 0
+    // error IncorrectMinAmount(uint amount);
+    // error IncorrectParticipants(uint limit);
+    // error MaxParticipantsReached(uint participants);
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        uint16 supply_
+    ) BatchNFT(name_, symbol_, supply_) {}
+
+    // start pre-order, only allowed by ower
+    function startPreOrder(uint256 amount, uint256 participants)
+        public
+        onlyOwner
+    {
+        // if (inPreOrder == true) revert PreOrderAlreadyStarted();
+        require(inPreOrder == false, "pre-order already started.");
```

Comment:
Shall we check if `amount` is smaller than `_totalSupply`?

Reply:
Ah I see, so we should check `participants` to be smaller than `_totalSupply`? Yes, `_totalSupply` is the number of NFT in supply, defined [here](https://github.com/thematters/matters-avatar/blob/develop/contracts/BatchNFT.sol#L16).